### DISCUSSION TITLE
Using Custom functions

### DISCUSSION THREAD

**Discussion by MPRDD** [sempare/sempare.boot.velocity.oss#216] [Q&A] [answered]
Hi.
I'm evaluating this incredible engine, to see if to use it for a rewite of direct debit xml generator I wrote years ago.

It's clear that the custom function mechanism works perfectly, but I think there is something in the concept of a context that I must be missing.

Any clarifying hints would be greatly appreciated.

--------------------
program Console;

{$APPTYPE CONSOLE}
{$R *.res}

uses
  System.SysUtils,
  System.IOUtils,
  Sempare.Template,
  Sempare.Template.Context;

type
  rPerson = record
    FirstName: string;
    LastName: string;
  end;

  TFunctions = class
  public
    class function FullName(const FirstName, LastName: string): string; static;
  end;

class function TFunctions.FullName(const FirstName, LastName: string): string;
  function ProperCase(aText: string): string;
  begin
    aText := trim(aText);
    result := uppercase(aText[1]) + lowercase(Copy(aText, 2));
  end;

begin
  result := ProperCase(FirstName) + ' ' + ProperCase(LastName);
end;

var
  me: rPerson;
  ctx: ITemplateContext;

const
  strMeTemplate = '<% FullName(_.FirstName, _.LastName) %>';

begin
  try
    me.FirstName := 'FIRSTNAME';
    me.LastName := 'LASTNAME';

    ctx := Template.Context([eoOptimiseTemplate, eoPrettyPrint]);
    ctx.Functions.AddFunctions(TFunctions);

    // case 1 OK
    writeln(Template.Eval(ctx, strMeTemplate, me));

    // ensure template file content is the same as strMeTemplate
    TFile.WriteAllText('.\templates\meTemplate.tpl', strMeTemplate);

    // case 2 KO ???
    writeln(Template.Eval(ctx, Template.Parse(TFile.ReadAllText('.\templates\meTemplate.tpl')), me));

    readln;
  except
    on E: Exception do
      Writeln(E.ClassName, ': ', E.Message);
  end;

end.

**Comment by darnocian** [comment, marked as answer]
The parser does attempt some validation on the function calls, so it also requires the context.

Change:
```
// case 2 KO ???
writeln(Template.Eval(ctx, Template.Parse(TFile.ReadAllText('.\templates\meTemplate.tpl')), me));
```
to
```
writeln(Template.Eval(ctx, Template.Parse(ctx, TFile.ReadAllText('.\templates\meTemplate.tpl')), me));
```

**Comment by darnocian** [comment]
Also, I'd recommend trying to use the template registry. It removes most of the boiler plate around loading from templates.

It has a single context that you can access **Template.Resolver.Context**.
When you use **Template.Resolve('template.tpl', data)**, it will resolve from a templates directory automatically. There is a **Template.Resolver.TemplateDir** that you can use to customise the location if required.

**Reply by darnocian** [reply]
FYI

https://github.com/sempare/sempare-delphi-template-engine/blob/main/docs/template-registry.md

https://dev.to/sempare/the-sempare-template-engine-advanced-template-registry-features-29o9

**Reply by MPRDD** [reply]
Thank you very much for your prompt and enlightening answer.
I'll certainly end up using the registry.
